### PR TITLE
Enable new CA1311 (ToLower/Upper culture) and CA1852 (seal internal/private types) rules

### PR DESCRIPTION
**Pull request by stephentoub**
CA1311 flagged a few issues, all addressed by using ToLowerInvariant/ToUpperInvariant.

CA1852 flagged a bunch that previous cleanups around sealing types missed or that are new since.  Sealing types then highlighted places where protected or virtual members were being exposed unnecessarily, so those were fixed, too.  Adding sealed to things also highlighted some discrepancies in the order of "unsafe sealed" keywords, where the vast majority in the repo were "sealed unsafe", so I fixed the few that weren't.

### CODE REVIEW

File: src/libraries/Common/src/System/Net/NTAuthentication.Managed.cs [hunk 1]
```diff
@@ -436,7 +436,7 @@ private static void makeNtlm2Hash(string domain, string userName, ReadOnlySpan<c
                 using (var hmac = IncrementalHash.CreateHMAC(HashAlgorithmName.MD5, pwHash))
                 {
                     // strangely, user is upper case, domain is not.
-                    byte[] blob = Encoding.Unicode.GetBytes(string.Concat(userName.ToUpper(), domain));
+                    byte[] blob = Encoding.Unicode.GetBytes(string.Concat(userName.ToUpperInvariant(), domain));
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Net/NTAuthentication.Managed.cs:439]: @wfurt, I assume this is correct?  I looked at the spec but it doesn't define "Uppercase(...)".

File: src/tasks/AppleAppBuilder/AppleAppBuilder.cs [hunk 2]
```diff
@@ -27,7 +27,7 @@ public string TargetOS
 
         set
         {
-            targetOS = value.ToLower();
+            targetOS = value.ToLowerInvariant();
```
Comment (stephentoub) [inline src/tasks/AppleAppBuilder/AppleAppBuilder.cs:30]: @radical, can you confirm the changes in this file and the one below are appropriate?

File: src/libraries/Common/src/System/Net/NTAuthentication.Managed.cs (hunk 1 above)
Comment (GrabYourPitchforks) [reply src/libraries/Common/src/System/Net/NTAuthentication.Managed.cs:439]: I'm not intimately familiar with NTLM, but in general, _ToUpperInvariant_ is the correct thing to do on Windows as far as usernames and other security identifiers are concerned.

File: src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ActivatorUtilitiesTests.cs [hunk 3]
```diff
@@ -391,15 +391,15 @@ public AbstractFoo()
             }
         }
 
-        private class Bar
+        private sealed class Bar
```
Comment (GrabYourPitchforks) [inline src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ActivatorUtilitiesTests.cs:394]: This directory looks like a test project - could sealing these types have subtle behavioral differences the test project didn't intend?

File: src/libraries/Microsoft.Extensions.Http/src/DefaultHttpClientFactory.cs [hunk 4]
```diff
@@ -194,13 +194,13 @@ internal void ExpiryTimer_Tick(object? state)
         }
 
         // Internal so it can be overridden in tests
```
Comment (GrabYourPitchforks) [inline src/libraries/Microsoft.Extensions.Http/src/DefaultHttpClientFactory.cs:196]: Does the unit test project have a friend relationship with the implementation project? This comment certainly implies sealing the type / methods would be a break.

File: src/mono/System.Private.CoreLib/src/System/__ComObject.cs [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 namespace System
 {
-    internal class __ComObject
+    internal sealed class __ComObject
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/mono/System.Private.CoreLib/src/System/__ComObject.cs:6]: For coreclr, yes.  For mono, I didn't think so, but I could be wrong... was hoping CI would tell me.

File: src/libraries/Common/src/System/Net/NTAuthentication.Managed.cs (hunk 1 above)
Comment (wfurt) [reply src/libraries/Common/src/System/Net/NTAuthentication.Managed.cs:439]: Would that work for international characters? I just did test on my local Windows 10 and I _can_ create user with localized name e.g. this does not count as special characters. I'm not sure how common that is. 
cc: @filipnavara  for visibility.

**Comment by GrabYourPitchforks** [comment]
Thanks for getting https://github.com/dotnet/roslyn-analyzers/pull/5972 up as well!

**Review by GrabYourPitchforks — APPROVED** [review]
Aside from the failing test (which means one of the sealed changes needs to be reverted?) - it all LGTM, including the _ToUpperInvariant/ToLowerInvariant_ portion.

File: src/libraries/Common/src/System/Net/NTAuthentication.Managed.cs (hunk 1 above)
Comment (filipnavara) [reply src/libraries/Common/src/System/Net/NTAuthentication.Managed.cs:439]: It's underspecified but I agree with @GrabYourPitchforks's assessment. `ToUpperInvariant` still works for national characters albeit with different rules and it makes sense that authentication would not depend on a specific locale.

**Pull request merged**